yrs/src/doc.rs

```diff
@@ -209,13 +293,75 @@ pub enum OffsetKind {
     Utf32,
 }
 
+pub trait Transact {
+    /// Creates a transaction used for all kind of block store operations.
+    /// Transaction cleanups & calling event handles happen when the transaction struct is dropped.
+    fn try_transact(&self) -> Result<Transaction, BorrowError>;
+
+    /// Creates a transaction used for all kind of block store operations.
+    /// Transaction cleanups & calling event handles happen when the transaction struct is dropped.
+    fn try_transact_mut(&self) -> Result<TransactionMut, BorrowMutError>;
+
+    /// Creates a transaction used for all kind of block store operations.
+    /// Transaction cleanups & calling event handles happen when the transaction struct is dropped.
+    fn transact(&self) -> Transaction {
+        self.try_transact().expect("cannot read document store contents because another read-write transaction is in progress")
+    }
+
+    /// Creates a transaction used for all kind of block store operations.
+    /// Transaction cleanups & calling event handles happen when the transaction struct is dropped.
+    fn transact_mut(&self) -> TransactionMut {
+        self.try_transact_mut()
+            .expect("only one read-write transaction can be active at the same time")
+    }
+}
+
+impl Transact for Doc {
+    fn try_transact(&self) -> Result<Transaction, BorrowError> {
+        Ok(Transaction::new(self.store.try_borrow()?))
+    }
+
+    fn try_transact_mut(&self) -> Result<TransactionMut, BorrowMutError> {
+        Ok(TransactionMut::new(self.store.try_borrow_mut()?))
+    }
+}
+
+impl Transact for Branch {
+    fn try_transact(&self) -> Result<Transaction, BorrowError> {
+        let store = self.store.as_ref().unwrap();
+        Ok(Transaction::new(store.try_borrow()?))
+    }
+
+    fn try_transact_mut(&self) -> Result<TransactionMut, BorrowMutError> {
+        let store = self.store.as_ref().unwrap();
+        Ok(TransactionMut::new(store.try_borrow_mut()?))
+    }
+}
+
+impl<T> Transact for T
+where
+    T: AsRef<Branch>,
```

Comment:
Now a new transaction can be created at any point in time either from the `Doc` level or straight from the y-types.